cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -731,7 +898,9 @@ class InputState {
 
   Status Push(const std::shared_ptr<arrow::RecordBatch>& rb) {
     if (rb->num_rows() > 0) {
-      queue_.Push(rb);
+      key_hasher_->Invalidate();  // batch changed - invalidate key hasher's cache
+      memo_.UpdateTime(GetTime(rb.get(), 0));  // time changed - update in MemoStore
```

Comment:
Can you explain why do we add the UpdateTime here. How was memostore time updated before?

Reply:
See [this post](https://github.com/apache/arrow/pull/34392#discussion_r1211671161). `UpdateTime` is synchronized, and before we didn't need this synchronization.